segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/ChunkedSegmentStorage.java

```diff
@@ -753,6 +754,30 @@ private void release(String... segmentNames) {
         }
     }
 
+    /**
+     * Executes the given Callable asynchronously and concurrently.
+     * It returns a CompletableFuture that will be completed with the result.
+     * The operations are assumed to be independent of other operations.
+     *
+     * @param operation    The Callable to execute.
+     * @param <R>       Return type of the operation.
+     * @param segmentNames The names of the Segments involved in this operation (for sequencing purposes).
+     * @return A CompletableFuture that, when completed, will contain the result of the operation.
+     * If the operation failed, it will contain the cause of the failure.
+     * */
+    private <R> CompletableFuture<R> executeParallel(Callable<CompletableFuture<R>> operation, String... segmentNames) {
+        return CompletableFuture.completedFuture(null).thenComposeAsync(v -> {
```

Comment:
Why so you need to start the computation this way?

Why not just submitting the operation to the thread pool?

The weird fact is that the operation returns a CompletableFuture so we may expect it to be already an asynchronous operation (but probably this is not always the case?)

Reply:
I didn't find any equivalent to CompletableFuture::ComposeAsync().
We definitely want to run this method and lambda supplied on storage thread pool and not on core thread pool.

Reply:
Use `CompletableFuture.runAsync` or `CompletableFuture.supplyAsync`, and inside that you execute your code. It's much simpler and more elegant.

Reply:
CompletableFuture.supplyAsync can't compose.

![image](https://user-images.githubusercontent.com/44757683/101112373-fca2f480-3591-11eb-9ed9-22c62fd48655.png)

```
Required type: CompletableFuture<R>
Provided: CompletableFuture<CompletableFuture<R>>

no instance(s) of type variable(s) exist so that CompletableFuture<R> conforms to R inference variable U has incompatible bounds: equality constraints: R lower bounds: CompletableFuture<R>
```

Reply:
OK, let's file a follow-up issue to clean this class up for the next release.